# Uso de filtros y contornos para la localización de placas en imágenes de coches
## 1.	En el siguiente código vamos a leer la imagen de un coche tanto a color como en niveles de gris, responde correctamente cuáles son los valores correctos que arroja alto 1 y ancho 1 y cuántos canales tiene:

In [13]:
import cv2 as cv

img01=cv.imread('Imagenes/cars96.png')
img02=cv.imread('Imagenes/cars96.png',cv.IMREAD_GRAYSCALE)
alto1, ancho1, canales1  = img01.shape
alto2, ancho2  = img02.shape

print("Alto1: ", alto1, "ancho1: ", ancho1, "canales: ", canales1)
print("Alto2: ", alto2, "ancho2: ", ancho2)

cv.imshow("img01",img01)
cv.imshow("img02",img02)
cv.waitKey(0)
cv.destroyAllWindows()

Alto1:  248 ancho1:  400 canales:  3
Alto2:  248 ancho2:  400


A. 248x400, 1 canal.  
**B. 248 x 400, 3 canales.✅**  
C. 320 x 460, 3 canales.  
D. 248 x 420, 1 canal.  

## 2.	El kernel aplicado a la imagen cars3.png es un filtro espacial____, y el resultado es la imagen:

In [14]:
import cv2 as cv
import numpy as np

img_gris=cv.imread('Imagenes/cars3.png',cv.IMREAD_GRAYSCALE)

kernel = np.array([[-1,-1,-1],
                   [-1, 8,-1],
                   [-1,-1,-1]])

img_ker = cv.filter2D(src=img_gris, ddepth=-1, kernel=kernel)

cv.imshow('Imagen con kernel', img_ker)
cv.waitKey(0)
cv.destroyAllWindows()
cv.imwrite('ResultadoImagenes/2-kernel.png',img_ker)

True

Pasa  
![Pasa ](ResultadoImagenes/kernel.png)

## 3.	El kernel aplicado a la imagen es un filtro ________ y se aplica principalmente para _________________. La imagen resultante es:

In [15]:
import cv2 as cv

img_gris = cv.imread('Imagenes/cars296.png',cv.IMREAD_GRAYSCALE)
kernel = cv.GaussianBlur(img_gris, (9, 9), 0)

cv.imshow("Imagen",img_gris)
cv.imshow("Imagen con filtro", kernel)
cv.waitKey(0)
cv.destroyAllWindows()
cv.imwrite('ResultadoImagenes/3-GaussianBlur.png',kernel)

True

![](ResultadoImagenes\3-GaussianBlur.png)

## 4.	El objetivo de la ecualización de imágenes es ________. El resultado al aplicar ecualización a la imagen coche11.jpg es:

In [16]:
# Cargar la imagen en color
import cv2 as cv
img = cv.imread('Imagenes/coche11.jpeg')

# Separa los canales de color de la imagen
b, g, r = cv.split(img)

# Ecualiza el histograma de cada canal de color
b_eq = cv.equalizeHist(b)
g_eq = cv.equalizeHist(g)
r_eq = cv.equalizeHist(r)

# Combina los canales de nuevo en una sola imagen 
img_eq = cv.merge((b_eq, g_eq, r_eq))
cv.imshow('Equalizada', img_eq)
cv.waitKey(0)
cv.destroyAllWindows
cv.imwrite('ResultadoImagenes/4-Ecualizacion.png',img_eq)

True

B. El objetivo de la ecualización es mejorar el contraste en una imagen, la imagen resultante es:  
![](ResultadoImagenes/ecualizacion.png)

## 5.	Para llevar a cabo el reconocimiento de placas es necesario antes localizarlas, existen una gran cantidad de condiciones que afectan una imagen como puede ser la iluminación, distancia y rotaciones, entre otros. Revisa las diferentes imágenes de coches que acompañan la actividad para que identifiques estos cambios. Para lograr el objetivo, hemos definido una metodología para encontrar placas candidatas en una imagen mediante técnicas tradicionales de procesamiento de imágenes, la metodología sugerida es la siguiente:

Paso 1. Leer imagen a color y transformar la imagen de color a gris.

In [17]:
import cv2 as cv
import numpy as np


img_color = cv.imread('Imagenes/coche16.jpeg')
img_gris = cv.imread('Imagenes/coche16.jpeg', cv.IMREAD_GRAYSCALE)

Paso 2. Suavizar la imagen (filtro gaussiano con kernel de 11x11).

In [18]:
blur = cv.GaussianBlur(img_gris, (11, 11), 0)
cv.imshow("Gaussiano", blur)
cv.imwrite('ResultadoImagenes/5.1-Gaussiano.png',blur)

True

Paso 3. Mejora de contraste (ajuste local de contraste adaptativo).

In [19]:
clahe = cv.createCLAHE(clipLimit=1.5, tileGridSize=(8,8))
equ01 = clahe.apply(blur)
cv.imshow("Ajuste de contraste", equ01)
cv.imwrite('ResultadoImagenes/5.2-Ajuste_de_contraste.png',equ01)

True

Paso 4. Binarizar la imagen (Otsu).

In [20]:
otsu_threshold, otsu01 = cv.threshold( equ01, 0, 255, cv.THRESH_OTSU)
cv.imshow("Imagen binarizada", otsu01)
cv.imwrite('ResultadoImagenes/5.3-Imagen_binarizada.png',otsu01)

True

Paso 5. Obtener bordes (CANNY).

In [21]:
edges = cv.Canny(otsu01, 150, 350)
cv.imshow("Canny", edges)
cv.imwrite('ResultadoImagenes/5.4-Canny.png',edges)

True

Paso 6. Guardar bordes.

In [22]:
contours, hierarchy = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
long_contours = [cnt for cnt in contours if (cv.arcLength(cnt, True) > 200
                                             and (cv.arcLength(cnt, True) < 700 ))]

Paso 7. Dibujar placas candidatas.

In [23]:
alto, ancho, canales = img_color.shape
for lonc in (long_contours):
    r = [x for [[x,y]] in lonc]
    s = [y for [[x,y]] in lonc]
    prop = (max(s)-min(s))-(max(r)-min(r))

    if ((max(s)>round(alto/2)) and prop<0):
        cv.rectangle(img_color,(min(r),min(s)),(max(r),max(s)),(0,255,0),3)

Paso 8. Cerrar ventanas y terminar.

In [24]:
cv. imshow("Imagen resultante", img_color)
cv.waitKey()
cv.destroyAllWindows()
cv.imwrite('ResultadoImagenes/5.5-Imagen_resultante.png',img_color)

True

Al ejecutar el código para la imagen coche16.jpeg, ¿cuál es la salida para los pasos 2, 3 , 4 y 5? Relaciona las columnas.

| Imagen | Paso |
|---------|-----|
|Imagen 1  
![Imagen binarizada paso 4](ResultadoImagenes/5.3-Imagen_binarizada.png)| Paso 4 |
|Imagen 2  
![Imagen binarizada paso 4](ResultadoImagenes/5.4-Canny.png)| Paso 5 |
|Imagen 3  
![Imagen binarizada paso 4](ResultadoImagenes/5.2-Ajuste_de_contraste.png)| Paso 3 |
|Imagen 4  
![Imagen binarizada paso 4](ResultadoImagenes/5.1-Gaussiano.png)| Paso 2 |

## 6.	Utilizando el siguiente algoritmo:
Paso 1. Leer imagen a color y transformar la imagen de color a gris.

In [2]:
import cv2 as cv
import numpy as np


img_color = cv.imread('Imagenes/coche16.jpeg')
img_gris = cv.imread('Imagenes/coche16.jpeg', cv.IMREAD_GRAYSCALE)

Paso 2. Suavizar la imagen (filtro gaussiano con kernel de 11x11).

In [3]:
blur = cv.GaussianBlur(img_gris, (11, 11), 0)
cv.imshow("Gausiano",blur)

Paso 3. Mejora de contraste (ajuste local de contraste adaptativo).

In [4]:
clahe = cv.createCLAHE(clipLimit=1.5, tileGridSize=(8,8))
equ01 = clahe.apply(blur)
cv.imshow("Ajuste de contraste", equ01)

Paso 4. Binarizar la imagen (Otsu).

In [5]:
otsu_threshold, otsu01 = cv.threshold( equ01, 0, 255, cv.THRESH_OTSU)
cv.imshow("Imagen binarizada", otsu01)

Paso 5. Obtener bordes (CANNY).

In [6]:
edges = cv.Canny(otsu01, 150, 350)
cv.imshow("Canny", edges)

Paso 6. Guardar bordes.

In [7]:
contours, hierarchy = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
long_contours = [cnt for cnt in contours if (cv.arcLength(cnt, True) > 200
                                             and (cv.arcLength(cnt, True) < 700 ))]

Paso 7. Dibujar placas candidatas.

In [8]:
alto, ancho, canales = img_color.shape
for lonc in (long_contours):
    r = [x for [[x,y]] in lonc]
    s = [y for [[x,y]] in lonc]
    prop = (max(s)-min(s))-(max(r)-min(r))

    if ((max(s)>round(alto/2)) and prop<0):
        cv.rectangle(img_color,(min(r),min(s)),(max(r),max(s)),(0,255,0),3)

Paso 8. Cerrar ventanas y terminar.

In [9]:
cv.imshow("Imagen resultante", img_color)
cv.waitKey()
cv.destroyAllWindows()

Si eliminamos el paso 2 de filtro gaussiano y lo aplicamos a la imagen coche14.jpeg, el resultado es:  
A. No detecta la placa.  
**B. Detecta más de cinco candidatos a placa.✅**   
C. Detecta exclusivamente la placa.  
D. Detecta tres candidatos de placa.  

### Desmostracion:

In [ ]:
# Paso 1. Leer imagen a color y transformar la imagen de color a gris.
import cv2 as cv
import numpy as np

img_color = cv.imread('Imagenes/coche14.jpeg')
img_gris = cv.imread('Imagenes/coche14.jpeg', cv.IMREAD_GRAYSCALE)
# Paso 2. Suavizar la imagen (filtro gaussiano con kernel de 11x11).
"""blur = cv.GaussianBlur(img_gris, (11, 11), 0)
cv.imshow("Gausiano",blur)"""
# Paso 3. Mejora de contraste (ajuste local de contraste adaptativo).
clahe = cv.createCLAHE(clipLimit=1.5, tileGridSize=(8,8))
equ01 = clahe.apply(img_gris)
cv.imshow("Ajuste de contraste", equ01)
# Paso 4. Binarizar la imagen (Otsu).
otsu_threshold, otsu01 = cv.threshold( equ01, 0, 255, cv.THRESH_OTSU)
cv.imshow("Imagen binarizada", otsu01)
# Paso 5. Obtener bordes (CANNY).
edges = cv.Canny(otsu01, 150, 350)
cv.imshow("Canny", edges)
# Paso 6. Guardar bordes.
contours, hierarchy = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
long_contours = [cnt for cnt in contours if (cv.arcLength(cnt, True) > 200
                                             and (cv.arcLength(cnt, True) < 700 ))]
# Paso 7. Dibujar placas candidatas.
alto, ancho, canales = img_color.shape
for lonc in (long_contours):
    r = [x for [[x,y]] in lonc]
    s = [y for [[x,y]] in lonc]
    prop = (max(s)-min(s))-(max(r)-min(r))

    if ((max(s)>round(alto/2)) and prop<0):
        cv.rectangle(img_color,(min(r),min(s)),(max(r),max(s)),(0,255,0),3)
#Paso 8. Cerrar ventanas y terminar.
cv.imshow("Imagen resultante", img_color)
cv.waitKey()
cv.destroyAllWindows()

True

| Sin paso 2 de filtro gaussiano | Con paso 2 de filtro gaussiano |
|--------------------------------|--------------------------------|
| ![](ResultadoImagenes/6.2-Imagen_res_Sin_Gaussiano.png) | ![](ResultadoImagenes/6.1-Imagen_res_Gaussiano.png) |

## 7. Utilizando el siguiente algoritmo Si incluimos todos los pasos y lo aplicamos a la imagen coche14.jpeg, el resultado es:
**A. Solo detecta la placa.**  
**B. Detecta tres candidatos a placa.**  
**<ins>C. Detecta dos candidatos a placa.</ins>✅**  
**D. No detecta ninguna placa.**  


Paso 1. Leer imagen a color y transformar la imagen de color a gris.
```python
1  import cv2 as cv
2  import numpy as np
3  
4  
5  img_color = cv.imread('Imagenes/coche14.jpeg')
6  img_gris = cv.imread('Imagenes/coche14.jpeg', cv.IMREAD_GRAYSCALE)
```
Paso 2. Suavizar la imagen (filtro gaussiano con kernel de 11x11).
```python
8  blur = cv.GaussianBlur(img_gris, (11, 11), 0)
9  cv.imshow("Gausiano",blur)
```
Paso 3. Mejora de contraste (ajuste local de contraste adaptativo).
```python
11  clahe = cv.createCLAHE(clipLimit=1.5, tileGridSize=(8,8))
12  equ01 = clahe.apply(blur)
13  cv.imshow("Ajuste de contraste", equ01)
```
Paso 4. Binarizar la imagen (Otsu).
```python
15  otsu_threshold, otsu01 = cv.threshold( equ01, 0, 255, cv.THRESH_OTSU)
16  cv.imshow("Imagen binarizada", otsu01)
```
Paso 5. Obtener bordes (CANNY).
```python
18  edges = cv.Canny(otsu01, 150, 350)
19  cv.imshow("Canny", edges)
```
Paso 6. Guardar bordes.
```python
21  contours, hierarchy = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
22  long_contours = [cnt for cnt in contours if (cv.arcLength(cnt, True) > 200
23                                               and (cv.arcLength(cnt, True) < 700 ))]
```
Paso 7. Dibujar placas candidatas.
```python
25  alto, ancho, canales = img_color.shape
26  for lonc in (long_contours):
27      r = [x for [[x,y]] in lonc]
28      s = [y for [[x,y]] in lonc]
29      prop = (max(s)-min(s))-(max(r)-min(r))
30  
31      if ((max(s)>round(alto/2)) and prop<0):
32          cv.rectangle(img_color,(min(r),min(s)),(max(r),max(s)),(0,255,0),3)
```
Paso 8. Cerrar ventanas y terminar.
```python
34  cv.imshow("Imagen resultante", img_color)
35  cv.waitKey()
36  cv.destroyAllWindows()
```

### Demostración


In [ ]:
# Paso 1. Leer imagen a color y transformar la imagen de color a gris.
import cv2 as cv
import numpy as np

img_color = cv.imread('Imagenes/coche14.jpeg')
img_gris = cv.imread('Imagenes/coche14.jpeg', cv.IMREAD_GRAYSCALE)
# Paso 2. Suavizar la imagen (filtro gaussiano con kernel de 11x11).
blur = cv.GaussianBlur(img_gris, (11, 11), 0)
cv.imshow("Gausiano",blur)
# Paso 3. Mejora de contraste (ajuste local de contraste adaptativo).
clahe = cv.createCLAHE(clipLimit=1.5, tileGridSize=(8,8))
equ01 = clahe.apply(blur)
cv.imshow("Ajuste de contraste", equ01)
# Paso 4. Binarizar la imagen (Otsu).
otsu_threshold, otsu01 = cv.threshold( equ01, 0, 255, cv.THRESH_OTSU)
cv.imshow("Imagen binarizada", otsu01)
# Paso 5. Obtener bordes (CANNY).
edges = cv.Canny(otsu01, 150, 350)
cv.imshow("Canny", edges)
# Paso 6. Guardar bordes.
contours, hierarchy = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
long_contours = [cnt for cnt in contours if (cv.arcLength(cnt, True) > 200
                                             and (cv.arcLength(cnt, True) < 700 ))]
# Paso 7. Dibujar placas candidatas.
alto, ancho, canales = img_color.shape
for lonc in (long_contours):
    r = [x for [[x,y]] in lonc]
    s = [y for [[x,y]] in lonc]
    prop = (max(s)-min(s))-(max(r)-min(r))

    if ((max(s)>round(alto/2)) and prop<0):
        cv.rectangle(img_color,(min(r),min(s)),(max(r),max(s)),(0,255,0),3)
#Paso 8. Cerrar ventanas y terminar.
cv.imshow("Imagen resultante", img_color)
cv.waitKey()
cv.destroyAllWindows()

True

##### Imagen resultante:  
![7 Imagen resultante](ResultadoImagenes/7-Imagen_resultante.png)